### PR TITLE
[ARCTIC-1167][core] Introduce `Trash` for restoring deleted files

### LINKED ISSUE #1167: [Feature]: Support move files to `Trash` when deleting files and restore files from `Trash` later
### Description

We should introduce a new feature to improve table security in the situation of deleting files, like snapshots expiration, and orphan files clean, making it easy to restore files when they are deleted by mistake.

### Use case/motivation

When files are deleted by mistake, whether they are metadata files or data files, it will cause serious problems for the iceberg table, like loading table errors or reading/writing errors.

And also, it's very difficult to recover the table from the abnormal state above; sometimes, restoring the files is the only way to recover the table.

### Describe the solution

1. We should introduce a new feature of Trash to restore deleted files
2. The `Trash` is an optional config for each Table
3. When deleting files of a table with `Trash` enabled, files are supposed to be moved to a specific directory, instead of deleted directly from the file system;  the `Trash` directory is also a config for each Table
4. We can restore deleted files from `Trash` when needed
5. The `Trash` directory is cleaned periodically with an expiration time by AMS

Get more details from the doc https://docs.google.com/document/d/1CYBb2u9KmhTcBVy-uUn_Lifu3osNnAgSkiD20LQhmko/edit#heading=h.74a2yp1p0du8

### Subtasks

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by wangtaohz**
## Why are the changes needed?

resolve #1167 

## Brief change log

  - Add API of TableTrashManager
  - add ArcticFileIOs and support building RecoverableArcticFileIO with a TableTrashManager inside
  - support setting a file pattern to decide whether a file should be moved to trash or deleted immediately
  - add TrashCleanService and a thread pool to AMS
  - refactor ScheduledTasks, support setting initial Delay, remove useless

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)
  https://docs.google.com/document/d/1CYBb2u9KmhTcBVy-uUn_Lifu3osNnAgSkiD20LQhmko/edit#heading=h.74a2yp1p0du8

### CODE REVIEW

File: site/docs/ch/configurations.md [hunk 1]
```diff
@@ -36,6 +36,10 @@ Self-optimizing 配置对 Iceberg format, Mixed streaming format 都会生效。
 | snapshot.base.keep.minutes                  | 720（12小时） | BaseStore 历史快照的保留时间                |
 | clean-orphan-file.enabled                   | false     | 是否开启孤儿文件自动清理                       |
 | clean-orphan-file.min-existing-time-minutes | 2880（2天）  | 存在时间超过 min-existing-time-minutes 未被引用的孤儿文件会被自动清理 |
+| table-trash.enabled                         | false     | 是否开启回收站功能                       |
```
Comment (zhoujinsong) [inline site/docs/ch/configurations.md:39]: As the table trash is available for mixed-format tables only, I'd recommend to move these configurations under the `Mixed streaming format` chapter.
And table trash is not only about table expiration, so we can add a sub chapter for this new feature.

File: core/src/test/java/com/netease/arctic/catalog/IcebergCatalogTest.java [hunk 2]
```diff
@@ -34,16 +41,73 @@ public IcebergCatalogTest() {
...
+    ArcticTable createTable = getCatalog().loadTable(TableTestHelpers.TEST_TABLE_ID);
+    Assert.assertEquals(false, PropertyUtil.propertyAsBoolean(createTable.properties(),
+        TableProperties.ENABLE_SELF_OPTIMIZING, TableProperties.ENABLE_SELF_OPTIMIZING_DEFAULT));
+    Assert.assertEquals(true, PropertyUtil.propertyAsBoolean(createTable.properties(),
+        TableProperties.ENABLE_TABLE_TRASH, TableProperties.ENABLE_TABLE_TRASH_DEFAULT));
+    Assert.assertFalse(createTable.io() instanceof RecoverableArcticFileIO);
+  }
+
+  @Test
+  public void testLoadTableWithNewCatalogProperties() throws TException {
```
Comment (zhoujinsong) [inline core/src/test/java/com/netease/arctic/catalog/IcebergCatalogTest.java:78]: If this test case is suitable for all the catalog types, I'd recommend to move it to `BaseCatalogTest`.

File: core/src/main/java/com/netease/arctic/catalog/BasicArcticCatalog.java [hunk 3]
```diff
@@ -608,10 +616,10 @@ protected KeyedTable createKeyedTable(TableMeta meta) {
         }
       });
 
-      BaseTable baseTable = new BasicKeyedTable.BasicInternalTable(tableIdentifier,
+      BaseTable baseTable = new BasicKeyedTable.BaseInternalTable(tableIdentifier,
```
[the remark this answers is not in the document]
Comment (wangtaohz) [reply core/src/main/java/com/netease/arctic/catalog/BasicArcticCatalog.java:624]: OK, I removed it.

File: site/docs/ch/configurations.md (hunk 1 above)
Comment (wangtaohz) [reply site/docs/ch/configurations.md]: I add a new sub-chapter `Trash 相关配置` to the `Mixed streaming format` chapter.

File: core/src/test/java/com/netease/arctic/catalog/IcebergCatalogTest.java (hunk 2 above)
Comment (wangtaohz) [reply core/src/test/java/com/netease/arctic/catalog/IcebergCatalogTest.java]: I moved both `testLoadTableWithNewCatalogProperties` and `testCreateTableWithCatalogTableProperties` to `BaseCatalogTest`.

**Pull request merged**